### PR TITLE
gh-117721: use PyMutex in `_thread.lock`

### PR DESCRIPTION
**Pull request by kumaraditya303**
Changes:
 - use `PyMutex` instead of `PyThread_type_lock`
 - as enhancement use `Py_TPFLAGS_MANAGED_WEAKREF`

* Issue: gh-117721

### CODE REVIEW

File: Modules/_threadmodule.c [hunk 1]
```diff
@@ -890,25 +877,20 @@ lock_repr(PyObject *op)
...
-    lockobject *self = (lockobject*)op;
-    if (_PyThread_at_fork_reinit(&self->lock_lock) < 0) {
-        PyErr_SetString(ThreadError, "failed to reinitialize lock at fork");
-        return NULL;
-    }
-
-    self->locked = 0;
-
+    lockobject *self = (lockobject *)op;
+    self->lock = (PyMutex){0};
```
Comment (colesbury) [inline Modules/_threadmodule.c]: This works fine, but you can also use `_PyMutex_at_fork_reinit(&self->lock)`

File: Modules/_threadmodule.c [hunk 2]
```diff
@@ -890,25 +877,20 @@ lock_repr(PyObject *op)
...
     Py_RETURN_NONE;
 }
 #endif  /* HAVE_FORK */
 
+
```
Comment (colesbury) [inline Modules/_threadmodule.c]: stray whitespace change

**Review by colesbury — APPROVED** [review]
LGTM. Minor comments below

File: Modules/_threadmodule.c (hunk 1 above)
Comment (kumaraditya303) [reply Modules/_threadmodule.c]: fixed

**Comment by colesbury** [comment]
@kumaraditya303, see https://github.com/python/cpython/issues/125058. We should check if this makes `threading.Lock.acquire()` interruptible on Windows. If so, it probably deserves a NEWS entry and possibly a documentation update.

**Comment by kumaraditya303** [comment]
> We should check if this makes threading.Lock.acquire() interruptible on Windows. If so, it probably deserves a NEWS entry and possibly a documentation update.

It doesn't, it is not interruptible.

**Comment by vstinner** [comment]
This change introduced a regression: https://github.com/python/cpython/issues/125451.

**Comment by vstinner** [comment]
> https://github.com/python/cpython/pull/125116 is a backport of this pull request to the [3.13 branch](https://github.com/python/cpython/tree/3.13).

IMO this change should not be backported to 3.13 in a bugfix release. It's too big, it can cause regression as you can see.

**Comment by colesbury** [comment]
I don't think there's a bug in `_thread.lock`, but I'm okay with reverting the backport

**Pull request merged**